multi-agent systems (#28)
MBrede/generative_ai: content/agent_interaction.qmd

---
toc-title: '![](../imgs/llm_as_a_judge.png){width=240px}<br> <h3>Agent interaction</h3>'
---

# Agent interactions

In this chapter, we want to introduce multi agent systems. As a starting point, we will talk about LLM as a judge. 

## LLM as a judge

Before we introduce the concept proper, let us first describe the problem it tries to solve: 

1. We generate text (be it natural language or structured output) using LLMs. 
2. The generated text is not always correct or appropriate for our use case.
3. We need a way to evaluate the quality of the generated text. 
4. To do this, we have to read it.
5. We don't have time for this.

The solution to this problem is, of course, to use an LLM to read and evaluate the text. This is only fair and proper, since it was an LLM that generated the text in the first place. The generated evaluation can then be used 

- to decide whether to accept or reject the generated text.
- to improve the model itself (e.g., for fine-tuning it on the generated text and its evaluation).
- to get an LLM to improve the text based on the evaluation.

> This approach is called LLM as a judge. It is a system that uses several calls to one or several LLMs to solve a problem. As such, it is an example of a multi-agent system. 

Reviewer: ```suggestion
This approach is called LLM as a judge. It is a system that uses several calls to one or several LLMs to solve a problem. As such, it can be implemented as a multi-agent system. 
```

This approach has a number of benefits as well as drawbacks. 

 - Benefits:
    - The evaluation can be very accurate and fast.
    - It is easy to implement.
    - It is easy to scale up the number of LLMs used for evaluation.
    - It is easy to use different LLMs for generation and evaluation.
    - It is easy to use different prompts for generation and evaluation.
  - Drawbacks:
    - The evaluation can be very expensive, since it requires several calls to the LLM.
    - The evaluation can be biased, since it is based on the LLMs' own evaluation of itself.
    - The evaluation can be subjective, since it is based on the LLMs' interpretation of the prompt.
    - The evaluation can be misleading, since it is based on the LLMs' interpretation of the generated text, which may not be the same as the human interpretation.

## A basic multi-agent system

>  Let us now look at a simple example of a multi-agent system. We will use the following scenario: We want to generate Anki flashcards from text.^[The following is loosely based on [@BuildingMultiAgentFramework2024], though I took the liberty to streamline and simplify the code a bit.]

Reviewer: ```suggestion
 Let us now look at a simple example of a multi-agent system. We will use the following scenario: We want to generate Anki flashcards from text.^[The following is loosely based on @BuildingMultiAgentFramework2024, though I took the liberty to streamline and simplify the code a bit.]
```

    class QACard(BaseModel):
    question: str
    answer: str
    extra: str -->

### Generator

Let us first implement the Anki card generator. It will take a text as input and return a card. A system prompt for the generator could look like this:

    You are an educational content creator specializing in Anki flashcard generation.
    Your task is to create one clear, concise flashcards following these guidelines:

    1. The card should focus on ONE specific concept
    2. The question should be clear and unambiguous
    3. The answer should be concise but complete
    4. Include relevant extra information in the extra field
    5. Follow the minimum information principle

    Format the card as:
    <card>
        <question>Your question here</question>
        <answer>Your answer here</answer>
        <extra>Additional context, examples, or explanations</extra>
    </card>

We will use llamaindex to implement the generator.

::: {.callout-note}
## 📝 Task

You can do it!

1. Open a notebook and connect it with a local LLM using LM Studio.

> 2. Initialize a generator agent without any tools. Do not use the ReAct agent this time, a simpler OpenAIAgent will do. 

Reviewer: funktioniert der OpenAIAgent ohne OpenAI?

Author: kann cih dir zeigen. habs von jeremy, ging bei mir erst auch nicht. der trick ist, openai von llamaindex zu impertoren, nicht von openai
